File: todo_weather_app/app/controllers/weather_controller.rb
require "net/http"
require "uri"
require "json"

class WeatherController < ApplicationController
  # Makes the rest calls to the weather API. Placed inside the controller to ensure api key is hidden
  def weather
    lat = params[:lat] || -33.91823523952793
    lon = params[:lon] || 18.42535499728242
    api_key = Rails.application.credentials.weather_api_key

    current_res = fetch_cached_data("temp", "http://api.weatherapi.com/v1/current.json?key=#{api_key}&q=#{lat},#{lon}")
    astronomy_res = fetch_cached_data("astro", "http://api.weatherapi.com/v1/astronomy.json?key=#{api_key}&q=#{lat},#{lon}")

    temp_c = JSON.parse(current_res)["current"]["temp_c"]
    sunrise = JSON.parse(astronomy_res)["astronomy"]["astro"]["sunrise"]
    sunset = JSON.parse(astronomy_res)["astronomy"]["astro"]["sunset"]

    render json: { temp_c: temp_c, sunrise: sunrise, sunset: sunset }
  end

  def fetch_cached_data(data_name, url)
    cache_key = "WEATHER_DATA_#{data_name}"
    Rails.cache.fetch(cache_key, expires_in: 1.hour) do
      uri = URI(url)
      Net::HTTP.get(uri)
    end
  end
end

File: todo_weather_app/app/javascript/controllers/weather_controller.js
import { Controller } from "@hotwired/stimulus"

// Connects to data-controller="weather"
export default class extends Controller {
  static targets = ["temperature", "sunrise", "sunset"];
  
  connect() {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(
        (position) => this.fetchWeatherData(position.coords.latitude, position.coords.longitude),
        () => this.fetchWeatherData()
      )
    } else {
      this.fetchWeatherData()
    }
  }

  async fetchWeatherData(lat, lon) {
    try {
      const weatherResponse = await fetch(`/weather?lat=${lat}&lon=${lon}`);
      const weatherData = await weatherResponse.json();

      if (!weatherResponse.ok) {
        this.temperatureTarget.textContent = "N/A"
        this.sunriseTarget.textContent = "N/A"
        this.sunsetTarget.textContent = "N/A"
      }
  
      this.temperatureTarget.textContent = `${weatherData.temp_c}°C`
      this.sunriseTarget.textContent = `${weatherData.sunrise}`
      this.sunsetTarget.textContent = `${weatherData.sunset}`
    } catch (error) {
      this.temperatureTarget.textContent = "N/A"
      this.sunriseTarget.textContent = "N/A"
      this.sunsetTarget.textContent = "N/A"
    }
    
  }
}
